### PR TITLE
Adding support for json responses from CLUSTER_SERVERS

### PR DESCRIPTION
**Pull request by clusterfudge**
Adding json support to RemoteFinder and RemoteReader. Some existing tests verified on inconsistent behavior, and this calls for thorough review.

### CODE REVIEW

File: webapp/tests/test_readers_remote.py [hunk 1]
```diff
@@ -179,6 +180,75 @@ def test_RemoteReader_fetch_multi(self, http_request):
...
 
+    @mock.patch('urllib3.PoolManager.request')
+    @mock.patch('django.conf.settings.CLUSTER_SERVERS', ['127.0.0.1', 'http://8.8.8.8/graphite?format=json&local=0'])
+    @mock.patch('django.conf.settings.INTRACLUSTER_HTTPS', False)
+    @mock.patch('django.conf.settings.REMOTE_STORE_USE_POST', False)
+    @mock.patch('django.conf.settings.FETCH_TIMEOUT', 10)
+    def test_RemoteReader_fetch_multi_json(self, http_request):
+        test_finders = RemoteFinder.factory()
+
+        # In test_RemoteReader_fetch_multi_json, we assert on the response start/end timestampes matching
```
Comment (clusterfudge) [inline webapp/tests/test_readers_remote.py]: @deniszh can you speak to this at all?

**Comment by clusterfudge** [comment]
Looks like I have some character encoding issues to fix in py3.x. Will tackle tomorrow.

**Comment by piotr1212** [comment]
I think you should use StringIO instead of BytesIO

**Comment by DanCech** [comment]
Is there a particular reason for wanting to use json over say msgpack?  I ask because the payloads are much larger with json so performance is likely to suffer quite a lot for larger queries.

File: webapp/graphite/readers/remote.py [hunk 2]
```diff
@@ -154,9 +152,46 @@ def _deserialize_buffer(byte_buffer, content_type):
...
+
+    if len(data) > 0 and len(data[0].get('datapoints', [])) > 1:
+      example_series_datapoints = data[0]['datapoints']
+      step = example_series_datapoints[1][1] - example_series_datapoints[0][1]
+      start_time = example_series_datapoints[0][1]
+      end_time = example_series_datapoints[-1][1]
+
+    return [
+      {
+        'pathExpression': str(series.get('target')),
```
[the remark this answers is not in the document]
Comment (clusterfudge) [reply webapp/graphite/readers/remote.py]: Thanks, Dan. In lieu of that (and my availability), I'm going to put this on hold and not attempt to get it into the 1.1.4 release this weekend. I'll reach out if I run into any issues, but off the top of my head I can honestly say I don't understand the purpose of pathExpression vs name. Can you clarify?

**Comment by clusterfudge** [comment]
@DanCech , based on some analysis that I ran recently and shared in this [gist](https://gist.github.com/clusterfudge/463ec623a9f10e450ff50efc03969bf4), json should be more performant on all metrics. Since the names of metrics aren't schema'd, there's really few savings in either space or compute (at least where I was testing on 2.7). If we wanted to take on an additional dependency, [ujson](https://pypi.org/project/ujson/) is an _amazing_ drop in replacement json library (the perf is bananas). 

Again, the above tests are local, somewhat contrived, and on python2.7, but the differences were pretty stark. The implementation of umsgpack embedded in graphite does not appear to be particularly performant, probably because it's a pure-python implementation (as opposed to something with native bindings).

**Comment by deniszh** [comment]
@clusterfudge : did you test proper msgpack module, like `u-msgpack-python` ?
Json just can't be faster then msgpack, that's completely counter-intuitive. :)

**Comment by deniszh** [comment]
```root@osboxes:~# python test.py
Serializer: test_json_buffer Payload Size: 80000
0.0690968036652
Serializer: test_json_stream Payload Size: 80000
0.0695509910583
Serializer: test_pickle_buffer Payload Size: 90006
0.590715885162
Serializer: test_pickle_stream Payload Size: 90006
18.5478489399
Serializer: test_umsgpack_buffer Payload Size: 90003
0.598016023636
Serializer: test_umsgpack_stream Payload Size: 90003
4.51842188835
Serializer: test_msgpacku_buffer Payload Size: 90003
0.0120220184326
Serializer: testumsgpacku_stream Payload Size: 90003
0.0128359794617```

**Comment by clusterfudge** [comment]
I used the message pack module that's embedded in graphite-web. Is that a fall back if the native implementation isn't installed?

**Comment by deniszh** [comment]
Sorry, not really getting what you mean, @clusterfudge 
Pure python version of modules is good for fallback, but performance sucks, obviously. 
If you need speed - just install an optimized version of the module.
`msgpacku` is the same drop-in replacement for `msgpack`, as `ujson` for `json`.

**Comment by deniszh** [comment]
```
Serializer: test_json_buffer Payload Size: 80000
0.0708241462708
Serializer: test_json_stream Payload Size: 80000
0.0717329978943
Serializer: test_ujson_buffer Payload Size: 70001
0.0287759304047
Serializer: test_ujson_stream Payload Size: 70001
0.0259568691254
```
```
Serializer: test_umsgpack_buffer Payload Size: 90003
0.641237020493
Serializer: test_umsgpack_stream Payload Size: 90003
4.80740189552
Serializer: test_msgpacku_buffer Payload Size: 90003
0.012020111084
Serializer: test_msgpacku_stream Payload Size: 90003
0.0133380889893
```

**Comment by deniszh** [comment]
> msgpacku is the same drop-in replacement for msgpack, as ujson for json.

Ooh, I was wrong, disregard that. `msgpack` is proper implementation, `umsgpack` is pure pyton, `msgpacku` is utf-8 wrapper.

File: webapp/graphite/readers/remote.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (DanCech) [reply webapp/graphite/readers/remote.py]: `pathExpression` is the actual query that the series matched, it's used as the key to get results back out of the `requestContext['prefetched']` data structure.

**Comment by deniszh** [comment]
Just did serialize benchmark one more time - looks like json is still fastest serializer from default Python distribution, both for 2 and 3. Msgpack is fastest when `msgpack` module is installed, with `cbor` on 2nd place, but CBOR also suffering from stream deserialization...

py2
```
Serializer: test_json_buffer Payload Size: 80000
0.111679077148
Serializer: test_json_stream Payload Size: 80000
0.0857601165771
Serializer: test_pickle_buffer Payload Size: 90006
0.968269109726
Serializer: test_pickle_stream Payload Size: 90006
19.1805100441
Serializer: test_msgpack_buffer Payload Size: 90003
0.0109009742737
Serializer: test_msgpack_stream Payload Size: 90003
0.0110709667206
Serializer: test_umsgpack_buffer Payload Size: 90003
0.940215110779
Serializer: test_umsgpack_stream Payload Size: 90003
4.4617190361
Serializer: test_cbor_buffer Payload Size: 90003
0.0151782035828
Serializer: test_cbor_stream Payload Size: 90003
3.91015386581
```
py3
```
Serializer: test_json_buffer Payload Size: 80000
0.11116572699999999
Serializer: test_json_stream Payload Size: 80000
0.10955180199999998
Serializer: test_pickle_buffer Payload Size: 90026
0.040406074000000014
Serializer: test_pickle_stream Payload Size: 90026
3.492091802
Serializer: test_msgpack_buffer Payload Size: 90003
0.018487617999999983
Serializer: test_msgpack_stream Payload Size: 90003
0.01901640300000018
Serializer: test_umsgpack_buffer Payload Size: 90003
1.0347651600000005
Serializer: test_umsgpack_stream Payload Size: 90003
4.593411390999999
Serializer: test_cbor_buffer Payload Size: 90003
0.022226440000000736
Serializer: test_cbor_stream Payload Size: 90003
3.972988451000001
```

**Comment by deniszh** [comment]
Well, IIRC cluster transport is an internal thing and was never published as a public API. Before Graphite 1.x only `picklie` was supported, which is python internal and schemaless, then @DanCech implement `msgpack` for speed and Metrictank interoperability...

**Comment by deniszh** [comment]
Are you still interested in implementing this, @clusterfudge ?

**Comment by clusterfudge** [comment]
I'm not; with the native python-msgpack library is installed, performance is perfectly acceptable.

**Pull request closed without merging**